rasa/core/featurizers.py

```diff
@@ -591,6 +579,96 @@ def _hash_example(states, action) -> int:
         frozen_actions = (action,)
         return hash((frozen_states, frozen_actions))
 
+    def _create_states_e2e(self, tracker):
+        import copy
+
+        def sublist(lst1, lst2):
+            len1 = len(lst1)
+            len2 = len(lst2)
+
+            if len1 >= len2:
+                return False
+
+            for i in range(0, len1):
+                if lst1[i] != lst2[i]:
+                    return False
+
+            return True
+
+        prev_tracker = None
+        states = []
+        for tr in tracker.generate_all_prior_trackers():
+            if prev_tracker:
+                if sublist(prev_tracker.applied_events(), tr.applied_events()):
+                    state = tr.applied_events()[len(prev_tracker.applied_events()) :]
+                    state_dict = {}
+                    for event in state:
+                        if isinstance(event, UserUttered):
+                            state_dict["user"] = Message(event.text)
+                        elif isinstance(event, ActionExecuted):
+                            state_dict["prev_action"] = Message(event.action_name, data={TEXT:event.action_name})
+                        slots = tr.current_slot_values()
+                        current_slots = {}
+                        for key, value in slots.items():
+                            if value:
+                                current_slots[key] = value
+                        state_dict["slots"] = current_slots
+                else:
+                    state_dict = {}
+            else:
+                # print(len(events[i-1]))
+                state_dict = {}
+            states.append(state_dict)
+            prev_tracker = copy.deepcopy(tr)
+        return states
+
+    def training_states_and_actions_e2e(
+        self, trackers: List[DialogueStateTracker], domain: Domain
+    ) -> Tuple[List[List[Optional[Dict[Text, float]]]], List[List[Text]]]:
+        trackers_as_states_e2e = []
+        trackers_as_actions_e2e = []
+        hashed_examples = set()
+        pbar = tqdm(
+            trackers, desc="Processed trackers e2e", disable=is_logging_disabled()
+        )
+        for tracker in pbar:
+            # tracker = trackers[0]
+            states = self._create_states_e2e(tracker)
+            idx = 0
+            for event in tracker.applied_events():
+                if isinstance(event, ActionExecuted):
+                    if not event.unpredictable:
+                        # only actions which can be
+                        # predicted at a stories start
+                        sliced_states = self.slice_state_history(
```

Comment:
this is the old type of slicing?

Reply:
with padding in the front

Reply:
no, in the slice_state_history we don't add padding.

Reply:
we do, these `None` that we add there is essentially padding

Reply:
Now, there are no `None`'s added in the slice history; I left the function but removed padding.